### PR TITLE
ARC-3008 Adding PDF scan for javascript with peepdf

### PR DESCRIPTION
**Pull request by jhalleeupgrade**

### CODE REVIEW

File: peepdf_wrapper_test.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import unittest
```
Comment (jhalleeupgrade) [inline peepdf_wrapper_test.py:1]: Real tests are coming!

**Review by gjacquetupgrade — APPROVED** [review]

File: peepdf_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        suspisious_elements = analysis_result['suspicious_elements']['actions']
+        for token in vulnerable_tokens:
+            if token in suspisious_elements:
+                return True
+    return False
+
+# peepdf is used from cmd-line with os.sys. Better to check the right version is present on the system.
+def check_peepdf_version_exists():
+    try:
+        pkg_resources.require(['peepdf==0.4.2'])
```
Comment (savilesupgrade) [inline peepdf_wrapper.py:46]: You'll need to update `requirements.txt` in the project too, as its referenced in the [Dockerfile](https://github.com/Credify/bucket-antivirus-function/blob/8229b184104cb11ea0ce28ba7b938bd322c20cbc/Dockerfile#L24)

Comment (jhalleeupgrade) [reply peepdf_wrapper.py:46]: Yes the dependency was added there

File: peepdf_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    try:
+        analysis_result = raw_json_string['peepdf_analysis']['advanced'][0]['version_info']
+    except KeyError as e:
+        print("Key in json couldn't be found.\nError: {0}".format(e))
+        sys.exit(1)
+
+    if 'suspicious_elements' in analysis_result:
+        suspisious_elements = analysis_result['suspicious_elements']['actions']
+        for token in vulnerable_tokens:
+            if suspisious_elements is not None and token in suspisious_elements:
```
Comment (savilesupgrade) [inline peepdf_wrapper.py:38]:
```suggestion
            if suspisious_elements and token in suspisious_elements:
```
More of a nit.
This should work, as an empty dict is False.

File: peepdf_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+from peepdf.main import main as peepdf_scanner
+
+def is_javascript_found(path):
+    # tokens used by peepdf in analysis result to indicate javascript was found
+    vulnerable_tokens = ['/JS', '/Javascript']
+
+    print("Starting peepdf scan of %s." % path)
+
+    # method peepdf_scanner() uses argsParser because it is a cmd line tool.
+    # So we need to pass arguments as it was called from the cmd-line.
```
Comment (savilesupgrade) [inline peepdf_wrapper.py:16]: Based on the peepdf module, we should be able to initiate this w/out invoking the cmdline.
We do lose the check if the file exists or not, that's all i can see from now.

(not really tested)
```
from peepdf.PDFCore import PDFParser

# https://github.com/jbremer/peepdf/blob/11dab2ead9bf344e6aa8cca7f82a17771350f1be/peepdf/main.py#L408
pdfParser = PDFParser()
ret, pdf = pdfParser.parse(path)  # other params are False defaults
statsDict = pdf.getStats()

version = '0.3'
revision = '275'
# https://github.com/jbremer/peepdf/blob/11dab2ead9bf344e6aa8cca7f82a17771350f1be/peepdf/main.py#L442
jsonReport = getPeepJSON(statsDict, _version, revision)
```

**Review by savilesupgrade — APPROVED** [review]
Minor suggestions / ideas but all else looks good.

**Comment by palacerteupgrade** [comment]
@dfosticupgrade do we have test cases to validate a change to the anti-virus check?

**Comment by jhalleeupgrade** [comment]
@palacerteupgrade @savilesupgrade This PR risks to change drastically (I might just create a new one). After testing with a considerate amount of "good" PDFs, I realised that peepdf seems to be a bit unstable.

I am currently working on adding custom yara definition files to the clamav scan (clamav supports adding custom yara definitions).

**Comment by palacerteupgrade** [comment]
Sounds good, no worries @jhalleeupgrade . It's weird because I assume we are not the first to do sanity checks on a PDF file.

**Pull request closed without merging**